jwkscache/cache.go

```diff
@@ -0,0 +1,270 @@
+	}
+
+	// Treat the location as the actual JWKS
+	// First, check if it's base64-encoded (remove trailing padding chars if present first)
+	locationJSON, err := base64.RawStdEncoding.DecodeString(strings.TrimRight(c.location, "="))
+	if err != nil {
+		// Assume it's already JSON, not encoded
+		locationJSON = []byte(c.location)
+	}
+
+	// Try decoding from JSON
+	c.jwks, err = jwk.Parse(locationJSON)
+	if err != nil {
+		return errors.New("failed to parse property 'location': not a URL, path to local file, or JSON value (optionally base64-encoded)")
+	}
+
+	return nil
+}
+
+func (c *JWKSCache) initJWKSFromURL(ctx context.Context, url string) error {
+	// Create the JWKS cache
+	cache := jwk.NewCache(ctx,
+		jwk.WithErrSink(httprc.ErrSinkFunc(func(err error) {
+			c.logger.Warnf("Error while refreshing JWKS cache: %v", err)
+		})),
+	)
+
+	// We also need to create a custom HTTP client (if we don't have one already) because otherwise there's no timeout.
+	if c.client == nil {
+		c.client = &http.Client{
+			Timeout: c.requestTimeout,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					MinVersion: tls.VersionTLS12,
+				},
+			},
+		}
+	}
+
+	// Register the cache
+	err := cache.Register(url,
+		jwk.WithMinRefreshInterval(c.minRefreshInterval),
+		jwk.WithHTTPClient(c.client),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to register JWKS cache: %w", err)
+	}
+
+	// Fetch the JWKS right away to start, so we can check it's valid and populate the cache
+	refreshCtx, refreshCancel := context.WithTimeout(ctx, c.requestTimeout)
+	_, err = cache.Refresh(refreshCtx, url)
+	refreshCancel()
+	if err != nil {
+		return fmt.Errorf("failed to fetch JWKS: %w", err)
+	}
+
+	c.jwks = jwk.NewCachedSet(cache, url)
+	return nil
+}
+
+func (c *JWKSCache) initJWKSFromFile(ctx context.Context, file string) error {
+	// Get the path to the folder containing the file
+	path := filepath.Dir(file)
+
+	// Start watching for changes in the filesystem
+	eventCh := make(chan struct{})
+	loaded := make(chan error, 1) // Needs to be buffered to prevent a goroutine leak
+	go func() {
+		watchErr := fswatcher.Watch(ctx, path, eventCh)
+		if watchErr != nil && !errors.Is(watchErr, context.Canceled) {
+			// Log errors only
+			c.logger.Errorf("Error while watching for changes to the local JWKS file: %v", watchErr)
+		}
+	}()
+	go func() {
+		var firstDone bool
+		for {
+			select {
+			case <-eventCh:
+				// When there's a change, reload the JWKS file
+				if firstDone {
+					c.logger.Debugf("Reloading JWKS file from disk")
+				} else {
+					c.logger.Debugf("Loading JWKS file from disk")
+				}
+				err := c.parseJWKSFile(file)
+				if !firstDone {
+					// The first time, signal that the initialization was complete and pass the error
+					loaded <- err
+					close(loaded)
+					firstDone = true
+				} else if err != nil {
+					// Log errors only
+					c.logger.Errorf("Error reading JWKS from disk: %v", err)
```

Comment:
How about load file when init and keep monitor about file changes? If so, we can remove `firstDone`.

Reply:
Doing it this way allows us to remove race conditions because there's only 1 goroutine that is loading the file. If we also load the file in the "main" goroutine, there's a possibility of a race condition if the file is changed on disk at the exact same time.

Reply:
At the same time, we also cannot start the watcher **after** having read the file in the "main" goroutine. There are no race conditions there, but in that case if the file changes on disk after we read it, but before the background goroutine is started, we may miss the updated file.

Reply:
Good explanation, please add comments here.